aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -69,7 +76,13 @@ pub(crate) enum DataRead<V> {
         #[derivative(PartialEq = "ignore", Debug = "ignore")] Arc<V>,
         #[derivative(PartialEq = "ignore", Debug = "ignore")] Option<Arc<MoveTypeLayout>>,
     ),
+    // Metadata and ResourceSize are insufficient to determine each other, but both
+    // can be determined from Versioned. When both are available, the information
+    // is stored in the MetadataAndResourceSize variant.
```

Comment:
this seems another optimization that's good to have? or can we consolidate Metadata/ResourceSize with this single one variant?

Reply:
Yes, the main issue was that previously we can upcast/downcast as there is a clear hierarchy, but size and metadata are kind of parallel to each other. So that's why we didn't originally implement size caching in CapturedReads. Now that I wanted to actually add it, the simplest way I found is to generalize the hierarchy and basically create an umbrella variant for size and metadata that can only be created internally (i.e. already cached size and now caching metadata, or vice versa)